configure.ac: configure script does not detect res_ninit correctly if resolv.h requires some other header files

On FreeBSD 14.0, after `./configure`,  `./config.log` shows:
```
...
configure:14284: checking for library containing res_ninit
configure:14310: cc -o conftest -O2 -pipe  -fstack-protector-strong -fno-strict-aliasing    -Wl,-rpath,/usr/local/lib -fstack-protector-strong  conftest.c  >&5
In file included from conftest.c:29:
/usr/include/resolv.h:158:14: error: array has incomplete element type 'struct sockaddr_in'
                nsaddr_list[MAXNS];     /*%< address of name server */
                           ^
/usr/include/resolv.h:157:9: note: forward declaration of 'struct sockaddr_in'
        struct sockaddr_in
               ^
/usr/include/resolv.h:172:18: error: field has incomplete type 'struct in_addr'
                struct in_addr  addr;
                                ^
/usr/include/resolv.h:172:10: note: forward declaration of 'struct in_addr'
                struct in_addr  addr;
                       ^
/usr/include/resolv.h:197:21: error: field has incomplete type 'struct sockaddr_in'
        struct sockaddr_in      sin;
                                ^
/usr/include/resolv.h:157:9: note: forward declaration of 'struct sockaddr_in'
        struct sockaddr_in
               ^
3 errors generated.
configure:14310: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "OpenDKIM"
| #define PACKAGE_TARNAME "opendkim"
| #define PACKAGE_VERSION "2.11.0"
| #define PACKAGE_STRING "OpenDKIM 2.11.0"
| #define PACKAGE_BUGREPORT "bugs@opendkim.org"
| #define PACKAGE_URL ""
| #define PACKAGE "opendkim"
| #define VERSION "2.11.0"
| #define HAVE_STDIO_H 1
| #define HAVE_STDLIB_H 1
| #define HAVE_STRING_H 1
| #define HAVE_INTTYPES_H 1
| #define HAVE_STDINT_H 1
| #define HAVE_STRINGS_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_UNISTD_H 1
| #define HAVE_SYS_TIME_H 1
| #define STDC_HEADERS 1
| #define HAVE_DLFCN_H 1
| #define LT_OBJDIR ".libs/"
| #define HAVE_PTHREAD_PRIO_INHERIT 1
| #define HAVE_PTHREAD 1
| #define HAVE_INET_PTON 1
| #define HAVE_INET_NTOP 1
| #define HAVE_GETADDRINFO 1
| /* end confdefs.h.  */
| #include <resolv.h>
| int
| main (void)
| {
| return res_ninit(NULL);
|   ;
|   return 0;
| }
...
configure:14465: checking for sys/types.h
configure:14465: result: yes
configure:14483: checking for netinet/in.h
configure:14483: result: yes
configure:14501: checking for arpa/nameser.h
configure:14501: result: yes
configure:14519: checking for netdb.h
configure:14519: result: yes
configure:14537: checking for resolv.h
configure:14537: result: yes
...
```

This is caused by the [configure.ac line 110-113](https://github.com/trusteddomainproject/OpenDKIM/blob/551ab3820476234def82eb2223ca6c7b45b45076/configure.ac#L110-L133) where use `resolv.h` in check for res_ninit before checking headers required by `resolv.h`.

With glibc, `resolv.h` itself contains all declarations for the function, so it is not affected.  However in all other platform which requires some other headers for using `resolv.h` but has `res_ninit()` function, the check would be failed.

Note:  In FreeBSD,  `res_setservers()` is also in libc as `__res_setservers` and `res_setservers` is a macro in `resolv.h`.  So it is need the same technique in the case of `res_ninit` if we want to detect it.  However there are PR #99 and issue #199 mentioned about `res_setservers`,  so I don't think we don't need to fix in the case `res_setservers`.

--- a/.gitignore
+++ b/.gitignore
@@ -21,6 +21,7 @@ build-config.h.in
 compile
 config.guess
 config.sub
+configure
 configure.scan
 configure.lineno
 aclocal.m4

--- a/configure.ac
+++ b/configure.ac
@@ -116,10 +116,26 @@ AC_SEARCH_LIBS(getaddrinfo, resolv,
 # it, it conflicts with AC_LANG_CALL's redeclaration. Hmm. I guess the
 # only thing for it is to include resolv.h, don't redeclare res_ninit(),
 # and use the proper type signature when calling it.
+
+# Before using resolv.h, we need to check headers required for it.
+AC_HEADER_RESOLV
+
 m4_rename([AC_LANG_CALL], [saved_AC_LANG_CALL])
-m4_define([AC_LANG_CALL], [AC_LANG_PROGRAM([#include <resolv.h>],
+m4_define([AC_LANG_CALL], [AC_LANG_PROGRAM([#ifdef HAVE_SYS_TYPES_H
+#  include <sys/types.h>
+#endif
+#ifdef HAVE_NETINET_IN_H
+#  include <netinet/in.h>
+#endif
+#ifdef HAVE_ARPA_NAMESER_H
+#  include <arpa/nameser.h>
+#endif
+#ifdef HAVE_NETDB_H
+#  include <netdb.h>
+#endif
+#include <resolv.h>],
                                            [return res_ninit(NULL);])])
-AC_SEARCH_LIBS(res_ninit, resolv,
+AC_SEARCH_LIBS(res_ninit, resolv bind,
                AC_DEFINE(HAVE_RES_NINIT, 1,
                          [Define to 1 if you have the `res_ninit()' function.]))
 m4_rename_force([saved_AC_LANG_CALL], [AC_LANG_CALL])
@@ -129,8 +145,6 @@ AC_SEARCH_LIBS(res_setservers, resolv bind,
 AC_SEARCH_LIBS(getopt_long, iberty,
                AC_DEFINE(HAVE_GETOPT_LONG, 1,
                          [Define to 1 if you have the `getopt_long()' function.]))
-
-AC_HEADER_RESOLV
 
 #
 # Check for types

